development/source-code-documentation.md

```diff
@@ -0,0 +1,187 @@
+TypeScript compatible [type checking](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html),
+if such type checking is supported by an editor or an IDE. For this we should follow the 
+[TypeScript JSDoc guidelines](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html).
+
+Documentation should follow the [ESLint JSDoc rules](https://eslint.org/docs/rules/valid-jsdoc).
+
+## Module
+An ES6 module should include a module name at the first line of the module file:
+```
+/** @module moduleName */
+```
+
+A module name should start from a lowercase letter in order to distinguish it from the class name.
+
+## Enums
+
+Define enums as:
+```
+/**
+ * Enum description
+ *
+ * @enum {string} */
+export const EnumType = {
+  VALUE_ONE: 'one',
+  VALUE_TWO: 'two'
+}
+```
+
+References:
+[TypeScript enum documentation guidelines](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#enum)
+
+
+## Classes
+
+```
+class C {
+  /**
+   * @param {number} data
+   */
+  constructor(data) {
+    // Property types can be inferred
+    this.name = "foo"
+
+    // or set explicitly
+    /** @type {string | null} */
+    this.title = null
+
+    // or simply annotated, if they're set elsewhere
+    /** @type {number} */
+    this.size
+    
+    // Properties can be declared as private
+    /** 
+     * @type {string}
+     * @private 
+     */
+    this._privateProp
+    
+    // Type info can be combined with description
+    /**
+     * A prop's description.
+     *
+     * @type {number}
+     */
+    this.propOne = 0
+    
+    // Description, type info, and visibility can be used together
+    /**
+     * A prop's description.
+     *
+     * @type {string}
+     * @private 
+     */
+    this._privatePropTwo = 0
+
+    this.initialize(data) // Error: initializer expects a string
+  }
+  
+  /**
+   * @param {string} s
+   */
+  initialize = function (s) {
+    this.size = s.length
+  };
+}
+
+let c = new C(0);
+```
+
+References:
+[TypeScript class documentation guidelines](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#classes)
+
+## Resolving types defined in other files
+
+Sometimes an IDE can not resolve types that are defined in other modules (other files). There are several solutions
+that can help with this. The choice of solution may depend on the IDE and its configuration.
+
+### Typedef in a target file
+
+One way to solve the issue is to use a combination of `typedef` with `import` in a target file:
+```
+import SomeClass from './some-class.js' /* @typedef {import('./some-class.js').SomeClass} SomeClass */
+
+/**
+* @param {SomeClass} value
+*/
+const f = (value) => {
+  value.prop = 'Some value'
+}
+```
+
+In this solution, webpack aliases can not be used in the `import` inside the `typedef` as an IDE will not be able to
+resolve them.
+
+Types can also be imported in place where they're referenced:
+```
+import SomeClass from './some-class.js'
```

Comment:
would it be helpful to standardize on one approach or the other?

Reply:
It kind of works for me in my IDE (IntelliJ), but not always and I'm still not sure I understand the mechanics. Sometimes it resolve classes without this statement. Sometimes it does not when the statement is present, but then starts to resolve it after a while.

I have now idea how it would work in Atom or Visual Studio Code. So I'm not sure I can recommend this approach for everything at the moment.

My intention was to throw this in and then gather some experience how it works in other environments. Then, based on that, we can come up with some common ground guidelines and update the documentation accordingly.

What do you think?

Reply:
ok